### PR TITLE
Json support

### PR DESCRIPTION
**Pull request by ofekshenawa**
- Add support for RedisJSON
- Add optional args and tests
- Add more tests
- Add more tests
- Add more tests and cleanups
- Add docstring
- update JSONArrIndex and matching texts
- update JSONArrTrim and matching texts
- update JSONGetWithArgs and matching texts
- Update JSONSetMode
- Updated JSONMSet
- Updated tests for JSONMSet
- Added stubbed (panicking) implementations of JSON.RESP AND JSON.DEBUG
- Pre-pull request tidy up
- Testing fixes

### CODE REVIEW

File: json.go [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+	JSONDel(ctx context.Context, key, path string) *IntCmd
+	JSONForget(ctx context.Context, key, path string) *IntCmd
+	JSONGet(ctx context.Context, key string, paths ...string) *JSONCmd
+	JSONGetWithArgs(ctx context.Context, key string, options *JSONGetArgs, paths ...string) *JSONCmd
+	JSONMerge(ctx context.Context, key, path string, value string) *StatusCmd
+	JSONMSetArgs(ctx context.Context, docs []JSONSetArgs) *StatusCmd
+	JSONMSet(ctx context.Context, params ...interface{}) *StatusCmd
+	JSONMGet(ctx context.Context, path string, keys ...string) *JSONSliceCmd
+	JSONNumIncrBy(ctx context.Context, key, path string, value float64) *JSONCmd
+	JSONNumMultBy(ctx context.Context, key, path string, value float64) *JSONCmd
```
Comment (chayim) [inline json.go:34]: Can we mark deprecated since it is (see commands.json) for NUMMULTBY

File: json.go [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+	JSONObjKeys(ctx context.Context, key, path string) *SliceCmd
+	JSONObjLen(ctx context.Context, key, path string) *IntPointerSliceCmd
+	JSONSet(ctx context.Context, key, path string, value interface{}) *StatusCmd
```
Comment (chayim) [inline json.go:37]: I don't see JSON.RESP. We implemented another command that's already deprecated. Either align + remove that one, or implement / panic appropriate this one.

File: json.go [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+		err := json.Unmarshal([]byte(cmd.val), &cmd.expanded)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return cmd.expanded, nil
+}
+
+func (cmd *JSONCmd) readReply(rd *proto.Reader) error {
```
Comment (chayim) [inline json.go:118]: @ofekshenawa can you have a look at spacing? In my browser the spacing seems completely off compared to other files. I don't know if this is just a github view.

**Review by RedisStack — APPROVED** [review]
Looks Good

File: json.go (hunk 3 above)
Comment (RedisStack) [reply json.go:118]: It is normal for me.

**Comment by chayim** [comment]
Please note - the above reply by @RedisStack is not a Redis account. This approval, and comment appears to be an impersonation attempt given the organization Redis maintains named [redis-stack])(https://github.com/redis-stack).

File: json.go (hunk 3 above)
Comment (ofekshenawa) [reply json.go:117]: It is normal for me.

**Comment by RedisStack** [comment]
> Please note - the above reply by @RedisStack is not a Redis account. This approval, and comment appears to be an impersonation attempt given the organization Redis maintains named [redis-stack])(https://github.com/redis-stack).

Sorry didn't mean to deceive people. I just really love Redis Stack

File: json.go [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+
+type JSONCmd struct {
+	baseCmd
+	val      string
+	expanded []interface{}
+}
+
+var _ Cmder = (*JSONCmd)(nil)
+
+func NewJSONCmd(ctx context.Context, args ...interface{}) *JSONCmd {
```
Comment (chayim) [inline json.go:68]: This appears to be internal - shouldn't this be lower cased, given the usage, only here? Is there a use case for this function - as opposed to *JSONCmd being returned by Get/GetWIthArgs/NumIncrBy?

**Review by chayim — APPROVED** [review]

**Pull request merged**